compiler/p/codegen/ControlFlowEvaluator.cpp

```diff
@@ -82,71 +82,423 @@ static void lookupScheme2(TR::Node *node, bool unbalanced, bool fromTableEval, T
+struct CompareInfo
    {
-   if (cg->comp()->target().is64Bit())
-      return TR::TreeEvaluator::iflucmpleEvaluator(node, cg);
+   CompareCondition cond;
+   TR::DataTypes type;
+   bool isUnsignedOrUnordered;
+
+   CompareInfo(CompareCondition cond, TR::DataTypes type, bool isUnsignedOrUnordered)
+      : cond(cond), type(type), isUnsignedOrUnordered(isUnsignedOrUnordered) {}
+   };
+
+bool is16BitSignedImmediate(int64_t value)
+   {
+   return value >= -0x8000 && value < 0x8000;
+   }
+
+bool is16BitUnsigedImmediate(uint64_t value)
+   {
+   return value < 0x10000;
+   }
+
+static TR::Register *evaluateAndSignExtend(TR::Node *node, bool extendInt32, TR::CodeGenerator *cg)
+   {
+   TR::Register *srcReg = cg->evaluate(node);
+
+   switch (node->getDataType().getDataType())
+      {
+      case TR::Int8:
+         {
+         TR::Register *trgReg = cg->allocateRegister();
+         generateTrg1Src1Instruction(cg, TR::InstOpCode::extsb, node, trgReg, srcReg);
+         return trgReg;
+         }
+
+      case TR::Int16:
+         {
+         TR::Register *trgReg = cg->allocateRegister();
+         generateTrg1Src1Instruction(cg, TR::InstOpCode::extsh, node, trgReg, srcReg);
+         return trgReg;
+         }
+
+      case TR::Int32:
+         if (extendInt32 && cg->comp()->target().is64Bit())
+            {
+            TR::Register *trgReg = cg->allocateRegister();
+            generateTrg1Src1Instruction(cg, TR::InstOpCode::extsw, node, trgReg, srcReg);
+            return trgReg;
+            }
+         else
+            {
+            return srcReg;
+            }
+
+      default:
+         return srcReg;
+      }
+   }
+
+static TR::Register *evaluateAndZeroExtend(TR::Node *node, bool extendInt32, TR::CodeGenerator *cg)
+   {
+   TR::Register *srcReg = cg->evaluate(node);
+
+   switch (node->getDataType().getDataType())
+      {
+      case TR::Int8:
+         {
+         TR::Register *trgReg = cg->allocateRegister();
+         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, trgReg, srcReg, 0, 0xff);
+         return trgReg;
+         }
+
+      case TR::Int16:
+         {
+         TR::Register *trgReg = cg->allocateRegister();
+         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, trgReg, srcReg, 0, 0xffff);
+         return trgReg;
+         }
+
+      case TR::Int32:
+         if (extendInt32 && cg->comp()->target().is64Bit())
+            {
+            TR::Register *trgReg = cg->allocateRegister();
+            generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, trgReg, srcReg, 0, 0xffffffff);
+            return trgReg;
+            }
+         else
+            {
+            return srcReg;
+            }
+
+      default:
+         return srcReg;
+      }
+   }
+
+static TR::Register *evaluateAndExtend(TR::Node *node, bool isUnsigned, bool extendInt32, TR::CodeGenerator *cg)
+   {
```

Comment:
this extendInt32 flag should be passed down to the node evaluator somehow for optimization. e.g.  deciding if we are going to use lwz vs. lwa.  it looks odd staying in this routine.

Reply:
I agree that information about whether zero-extension or sign-extension is preferred should be passed onto the node evaluator, and I actually created these helpers with the intention of later changing them to do this as part of some future work (so that the evaluators I've just refactored would not have to be changed to do this). I had a basic design for how such a thing could be done, but I haven't had the time to flesh it out yet. That being said, such work is outside the scope of this particular PR. I'll open an issue regarding this (with a proposed design) once this PR has been merged and these helpers actually exist in the codebase.